pkg/agent/cniserver/pod_monitor.go

```diff
@@ -0,0 +1,243 @@
+)
+
+var (
+	// retryInterval is the interval to re-install Pod OpenFlow entries if any error happened.
+	// Note, using a variable rather than constant for retryInterval because we may use a shorter time in the
+	// test code.
+	retryInterval = 5 * time.Second
+)
+
+type unreadyPodInfo struct {
+	podName        string
+	podNamespace   string
+	flowsInstalled bool
+	createTime     time.Time
+	ofPort         int32
+}
+
+type podIfaceMonitor struct {
+	podLister         corelisters.PodLister
+	kubeClient        clientset.Interface
+	eventBroadcaster  record.EventBroadcaster
+	record            record.EventRecorder
+	ifaceStore        interfacestore.InterfaceStore
+	ofClient          openflow.Client
+	podUpdateNotifier channel.Notifier
+
+	// unreadyInterfaces is a map to store the OVS ports which is waiting for the PortStatus from OpenFlow switch.
+	// The key in the map is the OVS port name, and its value is unreadyPodInfo.
+	// It is used only on Windows now.
+	unreadyInterfaces map[string]*unreadyPodInfo
+	// unreadyMapMutex is a mutex to guard map unreadyInterfaces.
+	unreadyMapMutex sync.Mutex
+	statusCh        chan *openflow15.PortStatus
+
+	queue workqueue.TypedDelayingInterface[string]
+}
+
+func (m *podIfaceMonitor) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting the monitor to wait for new OpenFlow ports")
+	m.eventBroadcaster.StartStructuredLogging(0)
+	m.eventBroadcaster.StartRecordingToSink(&v1.EventSinkImpl{
+		Interface: m.kubeClient.CoreV1().Events(""),
+	})
+	defer m.eventBroadcaster.Shutdown()
+
+	go wait.Until(m.worker, time.Second, stopCh)
+
+	go wait.Until(func() {
+		for {
+			select {
+			case status := <-m.statusCh:
+				klog.V(2).InfoS("Received PortStatus message", "message", status)
+				// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+				m.processPortStatusMessage(status)
+			}
+		}
+	}, time.Second, stopCh)
+}
+
+func (m *podIfaceMonitor) processPortStatusMessage(status *openflow15.PortStatus) {
+	// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+	if status.Desc.State == openflow15.PS_LIVE {
+		ovsPort := string(bytes.Trim(status.Desc.Name, "\x00"))
+		ofPort := status.Desc.PortNo
+		m.unreadyMapMutex.Lock()
+		defer m.unreadyMapMutex.Unlock()
+		podInfo, found := m.unreadyInterfaces[ovsPort]
+		if !found {
+			return
+		}
+		podInfo.ofPort = int32(ofPort)
+		m.queue.Add(ovsPort)
+	}
+}
+
+func (m *podIfaceMonitor) processNextWorkItem() bool {
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	defer m.queue.Done(key)
+
+	if err := m.updateUnreadyPod(key); err != nil {
+		// Put the item back on the workqueue to handle any transient errors.
+		m.queue.AddAfter(key, retryInterval)
+	}
+	return true
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the workqueue.
+func (m *podIfaceMonitor) worker() {
+	for m.processNextWorkItem() {
```

Comment:
Should we exit the loop if processNextWorkItem() returns false?

Reply:
no need, in `for-loop` it has the inject requirement that if the condition is true.